### PR TITLE
Upgrade Bazel version to 5.1.0.

### PR DESCRIPTION
**Pull request by ScottTodd**
Bazel 5.1.0 (https://blog.bazel.build/2022/03/24/bazel-5.1.html) added an experimental version of `cc_shared_library` behind the `experimental_cc_shared_library` flag. TensorFlow updated to Bazel 5.1.0 in https://github.com/tensorflow/tensorflow/commit/d01c0796aa461495ed0860b8cfe76db5b0f9914b and is using that flag and feature.

This updates IREE's Bazel version to 5.1.0 to match TensorFlow, and includes the newly required (thanks, TF) flag to our .bazelrc.

I did _not_ update the Docker image used for releases. We'll need something like https://github.com/stellaraccident/manylinux-bazel/commit/9937d5b5c1ca66136da1f25cf6f6470de20c9b58 for that.

### CODE REVIEW

**Comment by stellaraccident** [comment]
Fwiw, all of the funny business we had to do to make bazel work on that container is now gone. I think this could be reasonably scripted into IREE's docker containers now. Help on that would be great. Otherwise, open to pushing a new image when I'm back.

File: build_tools/bazel/iree.bazelrc [hunk 1]
```diff
@@ -15,6 +15,9 @@ build --experimental_guard_against_concurrent_changes
 build --compilation_mode=opt
 # Used in TensorFlow, so we have to enable it here as well.
 common --experimental_repo_remote_exec
+# TensorFlow uses cc_shared_library, which was added behind this experimental
+# flag in Bazel 5.1.0. Since we depend TensorFlow, we have to enable it too.
+build --experimental_cc_shared_library
```
Comment (GMNGeoffrey) [inline build_tools/bazel/iree.bazelrc:20]: We should be able to put this only in iree-tf.bazelrc

Comment (ScottTodd) [reply build_tools/bazel/iree.bazelrc:20]: Want to try that in this PR, or try moving `common --experimental_repo_remote_exec` and maybe some other flags there in a follow-up?

**Comment by ScottTodd** [comment]
> Fwiw, all of the funny business we had to do to make bazel work on that container is now gone. I think this could be reasonably scripted into IREE's docker containers now. Help on that would be great. Otherwise, open to pushing a new image when I'm back.

SG. I'd be up for helping with that. I vote we land this and bump LLVM+TF to keep the project moving, then fix the release builds later in the week (the path forward is pretty clear, and continuing to generate releases at the cost of slowing project development isn't worth it in this case, IMO).

File: build_tools/bazel/iree.bazelrc (hunk 1 above)
Comment (ScottTodd) [reply build_tools/bazel/iree.bazelrc]: I moved the new flag but left the old flags where they were. Hopefully this still passes our CI builds 🤞

**Review by GMNGeoffrey — APPROVED** [review]

**Comment by stellaraccident** [comment]
I pushed an update to 5.1.0 for the releases using the current images.

**Pull request merged**